android/src/main/java/io/grpc/android/AndroidChannelBuilder.java

```diff
@@ -0,0 +1,291 @@
+import io.grpc.ConnectivityState;
+import io.grpc.ExperimentalApi;
+import io.grpc.ForwardingChannelBuilder;
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+import io.grpc.MethodDescriptor;
+import io.grpc.internal.GrpcUtil;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * Builds a {@link ManagedChannel} that, when provided with a {@link Context}, will automatically
+ * monitor the Android device's network state to smoothly handle intermittent network failures.
+ *
+ * <p>Requires the Android ACCESS_NETWORK_STATE permission.
+ *
+ * @since 1.12.0
+ */
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/4056")
+public final class AndroidChannelBuilder extends ForwardingChannelBuilder<AndroidChannelBuilder> {
+
+  @Nullable
+  private static final Class<?> OKHTTP_CHANNEL_BUILDER_CLASS =
+      findClass("io.grpc.okhttp.OkHttpChannelBuilder");
+
+  private static final Class<?> findClass(String className) {
+    try {
+      return Class.forName(className);
+    } catch (ClassNotFoundException e) {
+      return null;
+    }
+  }
+
+  private final ManagedChannelBuilder delegateBuilder;
+
+  @Nullable private Context context;
+
+  public static final AndroidChannelBuilder forTarget(String target) {
+    return new AndroidChannelBuilder(target);
+  }
+
+  public static AndroidChannelBuilder forAddress(String name, int port) {
+    return forTarget(GrpcUtil.authorityFromHostAndPort(name, port));
+  }
+
+  private AndroidChannelBuilder(String target) {
+    if (OKHTTP_CHANNEL_BUILDER_CLASS == null) {
+      throw new UnsupportedOperationException("No ManagedChannelBuilder found on the classpath");
+    }
+    try {
+      delegateBuilder =
+          (ManagedChannelBuilder)
+              OKHTTP_CHANNEL_BUILDER_CLASS
+                  .getMethod("forTarget", String.class)
+                  .invoke(null, target);
+    } catch (Exception e) {
+      throw new RuntimeException("Failed to create ManagedChannelBuilder", e);
+    }
+  }
+
+  /** Enables automatic monitoring of the device's network state. */
+  public AndroidChannelBuilder context(Context context) {
+    this.context = context;
+    return this;
+  }
+
+  @Override
+  protected ManagedChannelBuilder<?> delegate() {
+    return delegateBuilder;
+  }
+
+  @Override
+  public ManagedChannel build() {
+    return new AndroidChannel(delegateBuilder.build(), context);
+  }
+
+  /**
+   * Wraps an OkHttp channel and handles invoking the appropriate methods (e.g., {@link
+   * ManagedChannel#resetConnectBackoff}) when the device network state changes.
+   */
+  @VisibleForTesting
+  static final class AndroidChannel extends ManagedChannel {
+
+    private final ManagedChannel delegate;
+
+    @Nullable private final Context context;
+    @Nullable private final ConnectivityManager connectivityManager;
+
+    @Nullable private DefaultNetworkCallback defaultNetworkCallback;
```

Comment:
I had sort of imagined these two fields would just be local variables, since the Runnables could just get a closure of them. This is fine too.

Reply:
Done